lib/connection_manager.go

```diff
@@ -386,133 +324,79 @@ func IPToNetAddr(ipStr string, addrMgr *addrmgr.AddrManager, params *DeSoParams)
 	return netAddr, nil
 }
 
-// ConnectPeer connects either an INBOUND or OUTBOUND peer. If Conn == nil,
-// then we will set up an OUTBOUND peer. Otherwise we will use the Conn to
-// create an INBOUND peer. If the connection is OUTBOUND and the persistentAddr
-// is set, then we will connect only to that addr. Otherwise, we will use
-// the addrmgr to randomly select addrs and create OUTBOUND connections
-// with them until we find a worthy peer.
-func (cmgr *ConnectionManager) ConnectPeer(conn net.Conn, persistentAddr *wire.NetAddress) {
-	// If we don't have a connection object then we will try and make an
-	// outbound connection to a peer to get one.
-	isOutbound := false
-	if conn == nil {
-		isOutbound = true
-	}
-	isPersistent := (persistentAddr != nil)
-	retryCount := 0
-	for {
-		// If the peer is persistent use exponential back off delay before retrying.
-		if isPersistent {
-			_delayRetry(retryCount, persistentAddr)
-		}
-		retryCount++
+func (cmgr *ConnectionManager) IsConnectedOutboundIpAddress(netAddr *wire.NetAddress) bool {
+	// Lock the address map since multiple threads will be trying to read
+	// and modify it at the same time.
+	cmgr.mtxConnectedOutboundAddrs.RLock()
+	defer cmgr.mtxConnectedOutboundAddrs.RUnlock()
+	return cmgr.connectedOutboundAddrs[addrmgr.NetAddressKey(netAddr)]
+}
 
-		// If this is an outbound peer, create an outbound connection.
-		if isOutbound {
-			conn = cmgr._getOutboundConn(persistentAddr)
-		}
+func (cmgr *ConnectionManager) IsAttemptedOutboundIpAddress(netAddr *wire.NetAddress) bool {
+	return cmgr.attemptedOutboundAddrs[addrmgr.NetAddressKey(netAddr)]
```

Comment:
Do we need a mutex around attemptedOutboundAddrs or nah?

Reply:
Yup, we do. Good call.